### DISCUSSION TITLE
RSA-based CMS signature not verified successfully when optional AlgorithmIdentifier parameters are omitted

### DISCUSSION THREAD

**Discussion by tbuchloh** [openssl/openssl#22095] [Q&A] [answered]
# Problem

The command does not lead to the expected outcome using openssl version 1.1.1+ up to 3.0.8:

```
openssl cms -verify -CAfile UN_CSCA.pem -in UN_r231.ml -inform DER -purpose any -out /dev/null
CMS Verification failure
40375E22107F0000:error:02000068:rsa routines:ossl_rsa_verify:bad signature:../crypto/rsa/rsa_sign.c:430:
40375E22107F0000:error:1C880004:Provider routines:rsa_verify:RSA lib:../providers/implementations/signature/rsa_sig.c:774:
40375E22107F0000:error:1700009E:CMS routines:CMS_SignerInfo_verify:verification failure:../crypto/cms/cms_sd.c:899:
```

# Background

The UN_r231.ml is a CSCA master list defined by ICAO Doc9303 Edition 8 Part 12 relying upon RFC 5652 (Cryptographic Message Syntax).

It is signed by a certificate chain "UN CSCA -> UN Master List Signer".

Both certificates are included in the `certificates` field.

# ASN.1 structure

```
    0:d=0  hl=5 l=580282 cons: SEQUENCE
    5:d=1  hl=2 l=   9 prim:  OBJECT            :pkcs7-signedData
   16:d=1  hl=5 l=580266 cons:  cont [ 0 ]
   21:d=2  hl=5 l=580261 cons:   SEQUENCE
   26:d=3  hl=2 l=   1 prim:    INTEGER           :03
   29:d=3  hl=2 l=  13 cons:    SET
   31:d=4  hl=2 l=  11 cons:     SEQUENCE
   33:d=5  hl=2 l=   9 prim:      OBJECT            :sha256
   44:d=3  hl=5 l=576733 cons:    SEQUENCE
   49:d=4  hl=2 l=   6 prim:     OBJECT            :2.23.136.1.1.2
   57:d=4  hl=5 l=576720 cons:     cont [ 0 ]
   62:d=5  hl=5 l=576715 prim:      OCTET STRING      [HEX DUMP]:308308CCC
   [...] eContent details left out
   B265F37A10ABCCACA41F4DA363C4EB882D399C963932177C64DB2E4D3623E7CCDF6CA5107F4300
576782:d=3  hl=4 l=3076 cons:    cont [ 0 ]
576786:d=4  hl=4 l=1436 cons:     SEQUENCE
576790:d=5  hl=4 l=1028 cons:      SEQUENCE
576794:d=6  hl=2 l=   3 cons:       cont [ 0 ]
576796:d=7  hl=2 l=   1 prim:        INTEGER           :02
576799:d=6  hl=2 l=   4 prim:       INTEGER           :599703A4
576805:d=6  hl=2 l=  13 cons:       SEQUENCE
576807:d=7  hl=2 l=   9 prim:        OBJECT            :sha256WithRSAEncryption
576818:d=7  hl=2 l=   0 prim:        NULL
[...] MLS certificate details left out
577795:d=8  hl=2 l=  25 cons:         SEQUENCE
577797:d=9  hl=2 l=   9 prim:          OBJECT            :1.2.840.113533.7.65.0
577808:d=9  hl=2 l=  12 prim:          OCTET STRING      [HEX DUMP]:300A1B0456382E32030204B0
577822:d=5  hl=2 l=  13 cons:      SEQUENCE
577824:d=6  hl=2 l=   9 prim:       OBJECT            :sha256WithRSAEncryption
577835:d=6  hl=2 l=   0 prim:       NULL
577837:d=5  hl=4 l= 385 prim:      BIT STRING
578226:d=4  hl=4 l=1632 cons:     SEQUENCE
578230:d=5  hl=4 l=1224 cons:      SEQUENCE
578234:d=6  hl=2 l=   3 cons:       cont [ 0 ]
578236:d=7  hl=2 l=   1 prim:        INTEGER           :02
578239:d=6  hl=2 l=   4 prim:       INTEGER           :5996E258
[...] CSCA certificate details left out
579458:d=5  hl=2 l=  13 cons:      SEQUENCE
579460:d=6  hl=2 l=   9 prim:       OBJECT            :sha256WithRSAEncryption
579471:d=6  hl=2 l=   0 prim:       NULL
579473:d=5  hl=4 l= 385 prim:      BIT STRING
579862:d=3  hl=4 l= 421 cons:    SET
579866:d=4  hl=4 l= 417 cons:     SEQUENCE
579870:d=5  hl=2 l=   1 prim:      INTEGER           :03
579873:d=5  hl=2 l=  20 prim:      cont [ 0 ]
579895:d=5  hl=2 l=  11 cons:      SEQUENCE
579897:d=6  hl=2 l=   9 prim:       OBJECT            :sha256
579908:d=5  hl=2 l= 102 cons:      cont [ 0 ]
579910:d=6  hl=2 l=  21 cons:       SEQUENCE
579912:d=7  hl=2 l=   9 prim:        OBJECT            :contentType
579923:d=7  hl=2 l=   8 cons:        SET
579925:d=8  hl=2 l=   6 prim:         OBJECT            :2.23.136.1.1.2
579933:d=6  hl=2 l=  28 cons:       SEQUENCE
579935:d=7  hl=2 l=   9 prim:        OBJECT            :signingTime
579946:d=7  hl=2 l=  15 cons:        SET
579948:d=8  hl=2 l=  13 prim:         UTCTIME           :230711140448Z
579963:d=6  hl=2 l=  47 cons:       SEQUENCE
579965:d=7  hl=2 l=   9 prim:        OBJECT            :messageDigest
579976:d=7  hl=2 l=  34 cons:        SET
579978:d=8  hl=2 l=  32 prim:         OCTET STRING      [HEX DUMP]:625E93663FCC56FE2A5DB18F43C49B0DE7D91D311AC10C2885D9DA8E8D955664
580012:d=5  hl=2 l=  13 cons:      SEQUENCE
580014:d=6  hl=2 l=   9 prim:       OBJECT            :sha256WithRSAEncryption
580025:d=6  hl=2 l=   0 prim:       NULL
580027:d=5  hl=4 l= 256 prim:      OCTET STRING      [HEX DUMP]:36FB45243885C762F97F173889C49653E8F9ED50F19443869033C44919975C45A432B14FA7A2B1E1A2F3450723FA9534A71E18A225E5BE2B6D60C52F70D0202CCB2392EE64705526A50BCCAA03ED5E3DC786B3FFEA2D18D58DB51A8CE883B4C95FD05D8F96A8092F77F02A0A5D32DC8E2EC9368BC8495BF3E70A7FE9BAE05C4918FE5BCA37D1C4FF0A77AEB60E0B7AF412BF12D5F4D0B9B2170828909C196DE53B53FDEA386EF1B4E4F4B11DE8D5A9846FC01C59656AC46EBD20C66D3BFBC46AA9C782996E0376BEAD66AE0FE6019A04D66C6C5B6A35217A764E7D96F596F0F53721A45E90FB1BC011CD9DB91D62A1EA3C60B73271E0E36FF11C662F65848BFC
```

# Observations

- O01: OpenSSL 1.0.2u verifies the CMS successfully.
- O02: BouncyCastle 1.76 verifies the CMS successfully.
- O03: Adding "-no_attr_verify" to the command line verifies the CMS successfully.
- O04: The recovered data from the signatureValue reveals, that AlgorithmIdentifier `sha256` does not encode `NULL` parameter.
  ```
  # get pubkey from MLS certificate
  openssl x509 -in "UN_MLS.pem" -noout -pubkey > pubkey.pem
  # extract signatureValue into file and strip off OCTET STRING header
  openssl asn1parse -in UN_r231.ml -inform DER -offset 580027 -out signature-raw.bin -noout
  tail --bytes=+5 signature-raw.bin > signature.bin
  # decrypt RSA-based signatureValue (TODO PKCS#1.5)
  openssl pkeyutl -verifyrecover -pubin -inkey pubkey.pem -in signature.bin > decrypted-signature.bin
  # parse decrypted signatureValue and note missing line with NULL parameter after ":sha256"
  openssl asn1parse -in decrypted-signature.bin -inform DER -i --dump
      0:d=0  hl=2 l=  47 cons: SEQUENCE
      2:d=1  hl=2 l=  11 cons:  SEQUENCE
      4:d=2  hl=2 l=   9 prim:   OBJECT            :sha256
     15:d=1  hl=2 l=  32 prim:  OCTET STRING      [HEX DUMP]:C13C4D1DC6FB089E456CB5298C17C311082921AB40FD5FB2A3BF57CB19739A5C
  ```
- O05: `ossl_rsa_verify()` in "rsa_sign.c" calls `encode_pkcs1()` that creates `encoded` (expected) value with explicitly encoded AlgorithmIdentifier parameter `05 00`.
  ![2023-09-13_14-09](https://github.com/openssl/openssl/assets/7712518/6eefc49b-1482-482a-bdef-61768ecbd886)
- O06: RFC 5754 (Using SHA2 Algorithms with CMS) states, that AlgorithmIdentifier parameters field is OPTIONAL:
  ```
   The AlgorithmIdentifier parameters field is OPTIONAL.
   Implementations MUST accept SHA2 AlgorithmIdentifiers with absent
   parameters.  Implementations MUST accept SHA2 AlgorithmIdentifiers
   with NULL parameters.  Implementations MUST generate SHA2
   AlgorithmIdentifiers with absent parameters.
  ```
- O07: RFC 3447 (PKCS#1: RSA Cryptography Specifications) appendix B.1 states, that parameters should be omitted, but if present shall have a NULL value.
  ```
   The parameters field associated with id-sha1, id-sha256, id-sha384,
   and id-sha512 should be omitted, but if present, shall have a value
   of type NULL.
  ```

# Expected result

openssl shall verify the CMS successfully, because it is compliant to RFC 5652, RFC 5754 and RFC 3447.

# Files

[files.zip](https://github.com/openssl/openssl/files/12597468/files.zip)

**Comment by t8m** [comment, marked as answer]
Thank you for your report. However this is not a bug and it will not be fixed.

The internal structure of the signed data in the RSA PKCS#1 v1.5 signature must be compliant with https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.2

This RFC explicitly requires the NULL parameter in the encoded data.

**Comment by t8m** [comment]
See also https://github.com/openssl/openssl/issues/17130

**Comment by tbuchloh** [comment]
Thank you for your explanation.

The reference you made, did not convince me, but reading the RFC a little bit further did.

> [A.2.2](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.2).  RSAES-PKCS-v1_5
> 
>    The object identifier rsaEncryption (see [Appendix A.1](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.1)) identifies the
>    RSAES-PKCS1-v1_5 encryption scheme.  The parameters field associated
>    with this OID in a value of type AlgorithmIdentifier SHALL have a
>    value of type NULL.  This is the same as in PKCS #1 v1.5.
> 
>        rsaEncryption    OBJECT IDENTIFIER ::= { pkcs-1 1 }

This OID is not used in the CMS at any place, but sha256 is, which lead to observations O06 and O07.

But when following the signature generation operation for sha256WithRSAEncryption we come to

> [8.2.1](https://www.rfc-editor.org/rfc/rfc8017#section-8.2.1).  Signature Generation Operation
> 
>    RSASSA-PKCS1-V1_5-SIGN (K, M)
> 
>    Input:
> 
>       K        signer's RSA private key
>       M        message to be signed, an octet string
> 
>    Output:
> 
>       S        signature, an octet string of length k, where k is the
>                length in octets of the RSA modulus n
> 
>    Errors:  "message too long"; "RSA modulus too short"
> 
>    Steps:
> 
>       1.  EMSA-PKCS1-v1_5 encoding: Apply the EMSA-PKCS1-v1_5 encoding
>           operation ([Section 9.2](https://www.rfc-editor.org/rfc/rfc8017#section-9.2)) to the message M to produce an encoded
>           message EM of length k octets:
> 
>              EM = EMSA-PKCS1-V1_5-ENCODE (M, k).
> 
>           If the encoding operation outputs "message too long", output
>           "message too long" and stop.  If the encoding operation
>           outputs "intended encoded message length too short", output
>           "RSA modulus too short" and stop.
> 

and then to

> [9.2](https://www.rfc-editor.org/rfc/rfc8017#section-9.2).  EMSA-PKCS1-v1_5
> 
>    This encoding method is deterministic and only has an encoding
>    operation.
> 
>    EMSA-PKCS1-v1_5-ENCODE (M, emLen)
> 
>    Option:
> 
>       Hash     hash function (hLen denotes the length in octets of
>                the hash function output)
> 
>    Input:
> 
>       M        message to be encoded
>       emLen    intended length in octets of the encoded message, at
>                least tLen + 11, where tLen is the octet length of the
>                Distinguished Encoding Rules (DER) encoding T of
>                a certain value computed during the encoding operation
> 
>    Output:
> 
>       EM       encoded message, an octet string of length emLen
> 
>    Errors:  "message too long"; "intended encoded message length too
>       short"
> 
>    Steps:
> 
>       1.  Apply the hash function to the message M to produce a hash
>           value H:
> 
>              H = Hash(M).
> 
>           If the hash function outputs "message too long", output
>           "message too long" and stop.
> 
>       2.  Encode the algorithm ID for the hash function and the hash
>           value into an ASN.1 value of type DigestInfo (see
>           [Appendix A.2.4](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.4)) with the DER, where the type DigestInfo has
>           the syntax
> 
>                DigestInfo ::= SEQUENCE {
>                    digestAlgorithm AlgorithmIdentifier,
>                    digest OCTET STRING
>                }
> 
>           The first field identifies the hash function and the second
>           contains the hash value.  Let T be the DER encoding of the
>           DigestInfo value (see the notes below), and let tLen be the
>           length in octets of T.

and then to

> [A.2.4](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.4).  RSASSA-PKCS-v1_5
> 
>    The object identifier for RSASSA-PKCS1-v1_5 SHALL be one of the
>    following.  The choice of OID depends on the choice of hash
>    algorithm: MD2, MD5, SHA-1, SHA-224, SHA-256, SHA-384, SHA-512,
>    SHA-512/224, or SHA-512/256.  Note that if either MD2 or MD5 is used,
>    then the OID is just as in PKCS #1 v1.5.  For each OID, the
>    parameters field associated with this OID in a value of type
>    AlgorithmIdentifier SHALL have a value of type NULL.  The OID should
>    be chosen in accordance with the following table:
> 
>          Hash algorithm   OID
>          ------------------------------------------------------------
>          MD2              md2WithRSAEncryption        ::= {pkcs-1 2}
>          MD5              md5WithRSAEncryption        ::= {pkcs-1 4}
>          SHA-1            sha1WithRSAEncryption       ::= {pkcs-1 5}
>          SHA-256          sha224WithRSAEncryption     ::= {pkcs-1 14}
>          SHA-256          sha256WithRSAEncryption     ::= {pkcs-1 11}
>          SHA-384          sha384WithRSAEncryption     ::= {pkcs-1 12}
>          SHA-512          sha512WithRSAEncryption     ::= {pkcs-1 13}
>          SHA-512/224      sha512-224WithRSAEncryption ::= {pkcs-1 15}
>          SHA-512/256      sha512-256WithRSAEncryption ::= {pkcs-1 16}
> 
>    The EMSA-PKCS1-v1_5 encoding method includes an ASN.1 value of type
>    DigestInfo, where the type DigestInfo has the syntax
> 
>        DigestInfo ::= SEQUENCE {
>            digestAlgorithm DigestAlgorithm,
>            digest OCTET STRING
>        }
> 
>    digestAlgorithm identifies the hash function and SHALL be an
>    algorithm ID with an OID in the set PKCS1-v1-5DigestAlgorithms.  For
>    a discussion of supported hash functions, see [Appendix B.1](https://www.rfc-editor.org/rfc/rfc8017#appendix-B.1).
> 
>        DigestAlgorithm ::= AlgorithmIdentifier {
>           {PKCS1-v1-5DigestAlgorithms}
>        }
> 
>        PKCS1-v1-5DigestAlgorithms    ALGORITHM-IDENTIFIER ::= {
>            { OID id-md2        PARAMETERS NULL }|
>            { OID id-md5        PARAMETERS NULL }|
>            { OID id-sha1       PARAMETERS NULL }|
>            { OID id-sha224     PARAMETERS NULL }|
>            { OID id-sha256     PARAMETERS NULL }|
>            { OID id-sha384     PARAMETERS NULL }|
>            { OID id-sha512     PARAMETERS NULL }|
>            { OID id-sha512-224 PARAMETERS NULL }|
>            { OID id-sha512-256 PARAMETERS NULL }
>        }

And this is the final proof for me, that openssl does it right (and BC + ML creator do not).

Sorry, for wasting your time.

**Reply by tzippy84** [reply]
Hi! I just came across your observation. Were you able to use openssl after all to verify your CMS signature? If yes, how if I may ask?
